### PR TITLE
Feat: Recurrent MAGPO

### PR DESCRIPTION
**Pull request by liyheng**
## What?
Implementation of recurrent MAGPO.
## Changes.
- Added implementation and configuration files for MAGPO.
- Updated `mava/networks/sable_network.py` and `mava/networks/utils/sable/decode.py`:  
  - Modified `SableNetwork` to return policy distributions for KL-distance calculation.
- Updated `mava/networks/heads.py`:  
  - Ensured that `scale` in `ContinuousActionHead` matches the shape of `loc` for correct reshaping in MAGPO.

### CODE REVIEW

File: mava/networks/utils/sable/decode.py [hunk 1]
```diff
@@ -79,8 +80,10 @@ def discrete_train_decoder_fn(
 
     distribution = distrax.Categorical(logits=masked_logits)
     action_log_prob = distribution.log_prob(action)
-
-    return action_log_prob, distribution.entropy()
+    if return_dist:
```
Comment (sash-a) [inline mava/networks/utils/sable/decode.py]: I would rather we always return the distribution and just ignore it where we don't need it, get's rid of all the `if`s

**Review by sash-a — CHANGES_REQUESTED** [review]
I'll go through the system file in detail tomorrow, but this looks really nice! Can you add a readme in the system folder with a short summary and link to the paper :pray:

File: mava/configs/network/rec_retention.yaml [hunk 2]
```diff
@@ -14,3 +14,30 @@ memory_config:
   timestep_chunk_size: ~ # Size of the chunk: calculated over timesteps dim.
   # For example a chunksize of 2 results in a sequence length of 2 * num_agents because there num_agents observations within a timestep
   # If unspecified, the rollout length is used as the chunk size which means that the entire rollout is computed in parallel during training.
+
+# ---Recurrent Structure Networks---
```
Comment (sash-a) [inline mava/configs/network/rec_retention.yaml]:
```suggestion
# ---Recurrent Structure Networks for MAGPO---
```

File: mava/systems/gpo/README.md
Comment (sash-a) [inline mava/systems/gpo/README.md:1]: :fire:

File: mava/systems/gpo/types.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    encoder: Array
+    decoder_self_retn: Array
+    decoder_cross_retn: Array
+
+class HiddenStates_all(NamedTuple):
+    """Hidden states for the encoder and decoder."""
+
+    sable_hidden_state: HiddenStates
+    policy_hidden_state: Array
```
Comment (sash-a) [inline mava/systems/gpo/types.py]: nit: can we rather call this `SableHiddenStates` and `HiddenStates`

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 4]
```diff
@@ -0,0 +1,804 @@
...
+        ----
+            learner_state (NamedTuple):
+                - params (FrozenDict): The current model parameters.
+                - opt_states (OptState): The current optimizer states.
+                - key (PRNGKey): The random number generator state.
+                - env_state (State): The environment state.
+                - last_timestep (TimeStep): The last timestep in the current trajectory.
+                - hstates (HiddenStates): The hidden state of the network.
+            _ (Any): The current metrics info.
+
```
Comment (sash-a) [inline mava/systems/gpo/anakin/rec_magpo.py]: nit
```suggestion
```

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 5]
```diff
@@ -0,0 +1,804 @@
...
+            last_done = done.repeat(env.num_agents).reshape(config.arch.num_envs, -1)
+            done = jnp.expand_dims(done, (1, 2, 3, 4))
+            sable_hstates = tree.map(lambda hs: jnp.where(done, jnp.zeros_like(hs), hs), sable_hstates)
+
+            prev_done = last_timestep.last().repeat(env.num_agents).reshape(num_envs, -1)
+            transition = Transition(
+                prev_done, action, value, timestep.reward, log_prob, last_timestep.observation, last_hstates
+            )
+            hstates = HiddenStates_all(sable_hstates, last_hstates.policy_hidden_state)
+            learner_state = LearnerState(params, opt_states, key, env_state, timestep, last_done, hstates)
```
Comment (sash-a) [inline mava/systems/gpo/anakin/rec_magpo.py]: last_done/prev_done mean the same thing to me, I just made it more explicit with curr_done/prev_done
```suggestion
            done = timestep.last()
            done = jnp.expand_dims(done, (1, 2, 3, 4))
            sable_hstates = tree.map(lambda hs: jnp.where(done, jnp.zeros_like(hs), hs), sable_hstates)

            curr_done = done.repeat(env.num_agents).reshape(config.arch.num_envs, -1)
            prev_done = last_timestep.last().repeat(env.num_agents).reshape(num_envs, -1)
            transition = Transition(
                prev_done, action, value, timestep.reward, log_prob, last_timestep.observation, last_hstates
            )
            hstates = HiddenStates_all(sable_hstates, last_hstates.policy_hidden_state)
            learner_state = LearnerState(params, opt_states, key, env_state, timestep, curr_done, hstates)
```

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 6]
```diff
@@ -0,0 +1,804 @@
...
+            _, advantages = jax.lax.scan(
+                _get_advantages,
+                (jnp.zeros_like(current_val), current_val, current_done),
+                traj_batch,
+                reverse=True,
+                unroll=16,
+            )
+            return advantages, advantages + traj_batch.value
+
+        advantages, targets = _calculate_gae(traj_batch, last_val, last_done)
```
Comment (sash-a) [inline mava/systems/gpo/anakin/rec_magpo.py]: Could you use our GAE util function :pray: 
Once you do just double check that performance is still as expected as it's quite easy to mess up the alignment between dones/obs with it. See [here](https://github.com/instadeepai/Mava/blob/develop/mava/systems/ppo/anakin/ff_ippo.py#L95-L97) for an example, but it looks to me like it is aligned correctly already

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 7]
```diff
@@ -0,0 +1,804 @@
...
+        dones=dones,
+        hstates=hstates,
+    )
+
+    return learn, actor_network, init_learner_state
+
+
+def run_experiment(_config: DictConfig) -> float:
+    """Runs experiment."""
+    _config.logger.system_name = "gpo"
```
Comment (sash-a) [inline mava/systems/gpo/anakin/rec_magpo.py]:
```suggestion
    _config.logger.system_name = "rec_magpo"
```

**Review by sash-a — CHANGES_REQUESTED** [review]
Looks really good, just a few minor comments from me!

Could you please run pre-commit on this as it seems our CI is failing because of linting errors

**Comment by liyheng** [comment]
I’ve run pre-commit and pushed the updated code. Thanks for pointing it out!

**Comment by sash-a** [comment]
Thanks @liyheng all the code looks great! Can I just see a performance graph that matches a result in your paper? Just to be sure everything is working :smile:

**Comment by liyheng** [comment]
I ran the code on three tasks, and the results look good.
3s5z_vs_3s6z
15x15-3p-5f
large-4ag-hard

One noteworthy difference between the implementation in our paper and this repository is that in the previous version we missed GAE normalization in _actor_loss_fn. This has been fixed here.
It doesn’t affect the performance much, but the optimal value of the weighting hyperparameter $\lambda$ might differ slightly from the one reported in our paper.

File: mava/configs/network/rec_retention.yaml [hunk 8]
```diff
@@ -14,3 +14,30 @@ memory_config:
   timestep_chunk_size: ~ # Size of the chunk: calculated over timesteps dim.
   # For example a chunksize of 2 results in a sequence length of 2 * num_agents because there num_agents observations within a timestep
   # If unspecified, the rollout length is used as the chunk size which means that the entire rollout is computed in parallel during training.
+
+# ---Recurrent Structure Networks for MAGPO---
```
Comment (RuanJohn) [inline mava/configs/network/rec_retention.yaml]: Could we please make a new network config for MAGPO which includes the RNN configs like you have here and then remove this from the `rec_retention.yaml` file?

File: mava/configs/system/gpo/rec_magpo.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+gamma: 0.99 # Discounting factor.
+gae_lambda: 0.95 # Lambda value for GAE computation.
+clip_eps: 0.2 # Clipping value for PPO updates and value function.
+ent_coef: 0.01 # Entropy regularisation term for loss function.
+vf_coef: 0.5 # Critic weight in
+max_grad_norm: 0.5 # Maximum norm of the gradients for a weight update.
+decay_learning_rates: False # Whether learning rates should be linearly decayed during training.
+target_kl: 0.01
+clip_gpo: 1.5
+alpha: 1
```
Comment (RuanJohn) [inline mava/configs/system/gpo/rec_magpo.yaml]: Could you please add a comment the explains what each of these parameters do like we do for the other config parameters?

File: mava/networks/heads.py [hunk 10]
```diff
@@ -100,7 +100,10 @@ def __call__(self, obs_embedding: chex.Array, action_mask: chex.Array) -> tfd.In
         del action_mask
         loc = self.mean(obs_embedding)
 
-        scale = self.log_std if self.independent_std else self.log_std(obs_embedding)
+        if self.independent_std:
+            scale = self.log_std * jnp.ones_like(loc)
+        else:
+            scale = self.log_std(obs_embedding)
```
Comment (RuanJohn) [inline mava/networks/heads.py:106]: Very nice!

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 11]
```diff
@@ -0,0 +1,816 @@
...
+                    # Reshape inputs
+                    obs, done, hidden = forward_reshape(
+                        (traj_batch.obs, traj_batch.done, traj_batch.hstates.policy_hidden_state),
+                        n_agents,
+                    )
+                    obs_and_done = (obs, done)
+
+                    _, actor_policy = actor_apply_fn(actor_params, hidden[0], obs_and_done)
+                    # Reshape output
+                    actor_policy = backward_reshape(actor_policy)
```
Comment (RuanJohn) [inline mava/systems/gpo/anakin/rec_magpo.py:324]: This is very cool. Could you please just add an explanation that we are not reshaping logits here, but the actual distribution?

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 12]
```diff
@@ -0,0 +1,816 @@
...
+                    log_prob_actor = actor_policy.log_prob(traj_batch.action)
+
+                    # Calculate kl loss
+                    if isinstance(actor_policy, IdentityTransformation):
+                        kl_loss = tfd.kl_divergence(
+                            lax.stop_gradient(policy), actor_policy.distribution
+                        ).mean()
+                    else:
+                        kl_loss = tfd.kl_divergence(lax.stop_gradient(policy), actor_policy).mean()
```
Comment (RuanJohn) [inline mava/systems/gpo/anakin/rec_magpo.py]: Could you please add a comment about why this check exists?

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 13]
```diff
@@ -0,0 +1,816 @@
...
+                        jnp.clip(
+                            log_prob - log_prob_actor,
+                            jnp.log(1 / config.system.clip_gpo),
+                            jnp.log(config.system.clip_gpo),
+                        )
+                        + log_prob_actor
+                        - traj_batch.log_prob
+                    )
+                    # Mask kl loss
+                    coe = jnp.select(
```
Comment (RuanJohn) [inline mava/systems/gpo/anakin/rec_magpo.py]: Could you please add a comment on what the coe variable is? Or change the naming?

File: mava/systems/gpo/anakin/rec_magpo.py [hunk 14]
```diff
@@ -0,0 +1,816 @@
...
+
+        def _update_epoch(update_state: Tuple, _: Any) -> Tuple:
+            """Update the network for a single epoch."""
+
+            def _update_minibatch(train_state: Tuple, batch_info: Tuple) -> Tuple:
+                """Update the network for a single minibatch."""
+                params, opt_states, key = train_state
+                traj_batch, advantages, targets, prev_hstates = batch_info
+
+                def _guider_loss_fn(
```
Comment (RuanJohn) [inline mava/systems/gpo/anakin/rec_magpo.py:207]: A general note on both the loss functions here. Could you please be more explicit with naming the variables where possible? For example, can variables relevant to the guider always start with something like `guider_` and similarly for the actor?

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you for this PR @liyheng! It is looking very good! I just have a few minor suggestions around variable naming and additional comments.

**Comment by sash-a** [comment]
One last thing @liyheng can you add your system to the table in the main readme :pray:

**Review by RuanJohn — APPROVED** [review]
Thank you for this contribution @liyheng!

**Review by sash-a — APPROVED** [review]
Thanks for all the work on this @liyheng! :fire:

**Comment by liyheng** [comment]
Thank you for your support and feedback!🚀

Additionally, I have implemented recurrent HAPPO and tested it across several domains. I noticed that in the issues, you previously tried HAPPO and found it didn’t perform well. In my tests, HAPPO also underperformed MAPPO in most scenarios.

Would you like me to create a PR for HAPPO and review the implementation?

**Comment by sash-a** [comment]
Thanks for the offer @liyheng. We decided not to include it in Mava because we don't think it's a worthwhile baseline, since our implementation was worse than MAPPO. 

If you are able to find a bug in your/our implementation such that it consistently outperforms MAPPO we would be more than happy to include it.
